src/main/java/tdl/record/sourcecode/snapshot/SnapshotRecorder.java

```diff
@@ -1,76 +1,154 @@
 package tdl.record.sourcecode.snapshot;
 
+import java.io.File;
+import java.io.FileFilter;
 import java.io.IOException;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.attribute.FileAttribute;
+import java.util.Date;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import org.apache.commons.io.FileUtils;
+import org.eclipse.jgit.api.Git;
+import org.eclipse.jgit.api.errors.GitAPIException;
 
 import tdl.record.sourcecode.content.SourceCodeProvider;
+import tdl.record.sourcecode.snapshot.helpers.DirectoryDiffUtils;
 
 public class SnapshotRecorder implements AutoCloseable {
 
     protected final SourceCodeProvider sourceCodeProvider;
 
-    private Path currentDirectorySnapshot;
+    private Git git;
 
-    private Path previousDirectorySnapshot;
+    private Path directory;
+
+    private Path gitDirectory;
 
     private int counter = 0;
-    private int keySnapshotPacing;
+
+    private final int keySnapshotPacing;
 
     public SnapshotRecorder(SourceCodeProvider sourceCodeProvider, int keySnapshotPacing) {
         this.sourceCodeProvider = sourceCodeProvider;
         this.keySnapshotPacing = keySnapshotPacing;
+        initTempDirectory();
+        initGitDirectory();
+    }
+
+    private void initTempDirectory() {
+        try {
+            File sysTmpDir = FileUtils.getTempDirectory();
+            directory = Files.createTempDirectory(
+                    sysTmpDir.toPath(),
+                    getClass().getSimpleName()
+            );
+        } catch (IOException ex) {
+            throw new RuntimeException(ex);
+        }
+    }
+
+    private void initGitDirectory() {
+        try {
+            File sysTmpDir = FileUtils.getTempDirectory();
+            gitDirectory = Files.createTempDirectory(
+                    sysTmpDir.toPath(),
+                    getClass().getSimpleName()
+            );
+            git = Git.init().setDirectory(gitDirectory.toFile()).call();
+            commitAllChanges();
+        } catch (IOException | GitAPIException ex) {
+            throw new RuntimeException(ex);
+        }
+    }
+
+    public Git getGit() {
+        return git;
+    }
+
+    public Path getGitDirectory() {
+        return gitDirectory;
+    }
+
+    public void syncToGitDirectory() throws IOException {
+        FileUtils.cleanDirectory(directory.toFile());
+        sourceCodeProvider.retrieveAndSaveTo(directory);
+        copyToGitDirectory();
```

Comment:
I was thinking about this and instead of doing this two step copy we could:
- remove all files from gitDirectory (except .git)
- sourceCodeProvider.retrieveAndSaveTo(gitDirectory)

This way you do not need to `removeDeletedFileInOriginal` and we only copy once.